File: Session_05_DynamicProgramming_Basics/multi_language_implementations/fibonacci_dp.js
const readline = require('readline');

function fibNaive(n) {
    if (n <= 1) return BigInt(n);
    return fibNaive(n - 1) + fibNaive(n - 2);
}

// Memoized helper
function fibMemoizeHelper(n, memo) {
    if (memo[n] !== -1n) return memo[n];
    if (n <= 1) {
        memo[n] = BigInt(n);
        return memo[n];
    }
    memo[n] = fibMemoizeHelper(n - 1, memo) + fibMemoizeHelper(n - 2, memo);
    return memo[n];
}

function fibMemoization(n) {
    const memo = new Array(n + 1).fill(-1n);
    return fibMemoizeHelper(n, memo);
}

function fibTabulation(n) {
    if (n <= 1) return BigInt(n);
    const table = new Array(n + 1);
    table[0] = 0n;
    table[1] = 1n;
    for (let i = 2; i <= n; i++) {
        table[i] = table[i - 1] + table[i - 2];
    }
    return table[n];
}

function main() {
    const n = 35;
    console.log(`Fib(${n}) Naive: ${fibNaive(n).toString()}`);
    console.log(`Fib(${n}) Memo: ${fibMemoization(n).toString()}`);
    console.log(`Fib(${n}) Tab: ${fibTabulation(n).toString()}`);
    console.log();

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question("Press Enter to exit...", () => {
        rl.close();
    });
}

main();

File: Session_05_DynamicProgramming_Basics/multi_language_implementations/fibonacci_dp.swift
import Foundation

func fibNaive(_ n: Int) -> Int64 {
    if n <= 1 {
        return Int64(n)
    }
    return fibNaive(n - 1) + fibNaive(n - 2)
}

func fibMemoizeHelper(_ n: Int, _ memo: inout [Int64]) -> Int64 {
    if memo[n] != -1 {
        return memo[n]
    }
    if n <= 1 {
        memo[n] = Int64(n)
        return memo[n]
    }
    memo[n] = fibMemoizeHelper(n - 1, &memo) + fibMemoizeHelper(n - 2, &memo)
    return memo[n]
}

func fibMemoization(_ n: Int) -> Int64 {
    var memo = [Int64](repeating: -1, count: n + 1)
    return fibMemoizeHelper(n, &memo)
}

func fibTabulation(_ n: Int) -> Int64 {
    if n <= 1 {
        return Int64(n)
    }
    var table = [Int64](repeating: 0, count: n + 1)
    table[0] = 0
    table[1] = 1
    for i in 2...n {
        table[i] = table[i - 1] + table[i - 2]
    }
    return table[n]
}

func main() {
    let n = 35
    print("Fib(\(n)) Naive: \(fibNaive(n))")
    print("Fib(\(n)) Memo: \(fibMemoization(n))")
    print("Fib(\(n)) Tab: \(fibTabulation(n))")
    print("\nPress Enter to exit...", terminator: "")
    fflush(stdout)
    let _ = readLine()
}

main()
